### PR TITLE
Add Windows and Mac CI stages

### PR DESCRIPTION
**Pull request by cmichi**
Follow-up to https://github.com/paritytech/ink/pull/1066.

The runs take quite long, I've enabled them only `master` for now.

### CODE REVIEW

**Comment by athei** [comment]
Those jobs are almost the same. Couldn't they be unified to one and just using the platform matrix feature?

File: .github/workflows/macos.yml [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+      - name: Build examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
+        run: |
+          for example in examples/*/; do
+              echo "Building example: $example";
+              cargo contract build --manifest-path=$example/Cargo.toml;
+          done
+          cd examples/delegator/ && ./build-all.sh
+
+      # For the moment we don't test the examples on MacOS, since it takes a long
```
Comment (HCastano) [inline .github/workflows/macos.yml:69]: I'd open an issue for this and link to that here instead of leaving commented out code

**Review by HCastano — COMMENTED** [review]
Same comment as Alex wrt. to sharing code

File: .github/workflows/test-examples.yml [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+jobs:
+  check:
+    name: build-examples
+    strategy:
+      matrix:
+        platform:
+          - macos-latest
+          - windows-latest
+        toolchain:
+          - nightly
```
Comment (athei) [inline .github/workflows/test-examples.yml:26]:
```suggestion
        toolchain:
          - nightly
        job:
          - build
          - test
```

File: .github/workflows/test-examples.yml [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+           foreach ($contract in $delegator_subcontracts) {
+               echo "Testing delegator contract: $contract";
+               cargo test --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
+           }
+           foreach ($example in Get-ChildItem  -Directory "examples\*") {
+               echo "Testing example: $example";
+               cargo test --verbose --manifest-path=$example/Cargo.toml;
+           }
+
+      - name: Test examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
```
Comment (athei) [inline .github/workflows/test-examples.yml:78]:
```suggestion
      - name: ${{ matrix.job }} examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
```

File: .github/workflows/test-examples.yml [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            command: install
+            args: --git https://github.com/paritytech/cargo-contract.git
+
+      - name: Output versions
+        run: |
+            wasm-opt --version
+            cargo -vV
+            cargo contract --version
+
+      - name: Test examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
```
Comment (athei) [inline .github/workflows/test-examples.yml:65]:
```suggestion
      - name: ${{ matrix.job }} examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
```

File: .github/workflows/test-examples.yml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        run: |
+           $delegator_subcontracts = "accumulator","adder","subber"
+           foreach ($contract in $delegator_subcontracts) {
+               echo "Testing delegator contract: $contract";
+               cargo test --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
+           }
+           foreach ($example in Get-ChildItem  -Directory "examples\*") {
+               echo "Testing example: $example";
+               cargo test --verbose --manifest-path=$example/Cargo.toml;
+           }
```
Comment (athei) [inline .github/workflows/test-examples.yml:76]: Also swapped the position of the loop to be consistent with macOS:

```suggestion
           $delegator_subcontracts = "accumulator","adder","subber"
           foreach ($contract in $delegator_subcontracts) {
               echo "Processing delegator contract: $contract";
               cargo ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
           }     
          foreach ($example in Get-ChildItem  -Directory "examples\*") {
               echo "Processing example: $example";
               cargo ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
           }
            
```

File: .github/workflows/test-examples.yml [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        if: runner.os == 'macOS'
+        run: |
+           for example in examples/*/; do
+               echo "Testing example: $example";
+               cargo test --verbose --manifest-path=$example/Cargo.toml;
+           done
+           for contract in ${DELEGATOR_SUBCONTRACTS}; do
+               echo "Testing delegator contract: $contract";
+               cargo test --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
+           done
```
Comment (athei) [inline .github/workflows/test-examples.yml:88]:
```suggestion
           for example in examples/*/; do
               echo "Processing example: $example";
               cargo ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
           done
           for contract in ${DELEGATOR_SUBCONTRACTS}; do
               echo "Processing delegator contract: $contract";
               cargo ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
           done
```

**Review by athei — COMMENTED** [review]
We can also use the matrix feature to remove the duplication between test and build I think. See my suggestions.

**Comment by cmichi** [comment]
@athei Thanks for the suggestions, the build and test is sequential now though and no longer parallel.

**Comment by cmichi** [comment]
Individual durations currently:

Platform | What? | How long?
:---|:---|:---
Windows | Build examples | 1h 7min
MacOS | Build examples | 1h 23min
Windows | Test examples | 33min
MacOS | Test examples | 35min

**Review by athei — APPROVED** [review]

File: .github/workflows/test-examples.yml [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            cargo -vV
+            cargo contract --version
+
+      - name: ${{ matrix.job }} examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
+        if: runner.os == 'Windows'
+        run: |
+           $delegator_subcontracts = "accumulator","adder","subber"
+           foreach ($contract in $delegator_subcontracts) {
+               echo "Processing delegator contract: $contract";
+               cargo ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
```
Comment (athei) [inline .github/workflows/test-examples.yml:70]:
```suggestion
               cargo contract ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
```

File: .github/workflows/test-examples.yml [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+           }
+           foreach ($example in Get-ChildItem  -Directory "examples\*") {
+               echo "Processing example: $example";
+               cargo ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
```
Comment (athei) [inline .github/workflows/test-examples.yml:74]:
```suggestion
               cargo contract ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
```

File: .github/workflows/test-examples.yml [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+           }
+
+      - name: ${{ matrix.job }} examples on ${{ matrix.platform }}-${{ matrix.toolchain }}
+        if: runner.os == 'macOS'
+        run: |
+           for example in examples/*/; do
+               echo "Processing example: $example";
+               cargo ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
```
Comment (athei) [inline .github/workflows/test-examples.yml:82]:
```suggestion
               cargo contract ${{ matrix.job }} --verbose --manifest-path=$example/Cargo.toml;
```

File: .github/workflows/test-examples.yml [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+           done
+           for contract in ${DELEGATOR_SUBCONTRACTS}; do
+               echo "Processing delegator contract: $contract";
+               cargo ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
```
Comment (athei) [inline .github/workflows/test-examples.yml:86]:
```suggestion
               cargo contract ${{ matrix.job }} --verbose --manifest-path examples/delegator/${contract}/Cargo.toml;
```

**Pull request merged**